File: 백준/Gold/15686. 치킨 배달/치킨 배달.java

import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.OutputStreamWriter;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashSet;
import java.util.List;
import java.util.Set;
import java.util.StringTokenizer;

class Point{
	int row,col;
	
	Point(int row,int col){
		this.row = row;
		this.col = col;
	}

	public int getRow() {
		return row;
	}

	public int getCol() {
		return col;
	}

	@Override
	public String toString() {
		return "Pair [row=" + row + ", col=" + col + "]";
	}
}
public class Main {
	
	static int N,M;
	static int[][] city;
	static List<Point> chicken; 
	static List<Point> house;
	static boolean chicken_chk[];
	static int answer=Integer.MAX_VALUE;
	
	static void backTracking(int start,int cnt) {
		
		if(cnt==M) { // M개의 치킨 집을 선택하면 도시의 치킨 거리 계산 시작
			int result=0;
			for(int p=0;p<house.size();p++) {
				int min_distance = Integer.MAX_VALUE;
				for(int q=0;q<chicken.size();q++) {
					if(chicken_chk[q]) {
						int now_distance = Math.abs(house.get(p).getRow()-chicken.get(q).getRow()) + Math.abs(house.get(p).getCol()-chicken.get(q).getCol());
						
						min_distance = Math.min(min_distance, now_distance);
						
						if(min_distance==1)
							break;
					}
				}
				result+=min_distance;
			}
			
			answer = Math.min(answer, result);
			
			return;
		}
		
		for(int i=start;i<chicken.size();i++) { // 치킨집의 모든 경우를 백트래킹으로 탐색함
			chicken_chk[i]=true;
			backTracking(i+1,cnt+1);
			chicken_chk[i]=false;
		}
	}
	
	public static void main(String[] args) throws IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st = new StringTokenizer(br.readLine());
		
		chicken = new ArrayList<>();
		house = new ArrayList<>();
		N = Integer.parseInt(st.nextToken());
		M = Integer.parseInt(st.nextToken());
		city = new int[N][N];
		
		for(int i=0;i<N;i++) {
			st = new StringTokenizer(br.readLine());
			for(int j=0;j<N;j++) {
				city[i][j] = Integer.parseInt(st.nextToken());
				if(city[i][j]==2) { // 치킨 집 이면 chicken List에 좌표값을 저장
					chicken.add(new Point(i,j));
				}else if(city[i][j]==1) {
					house.add(new Point(i,j));
				}
			}
		}
		chicken_chk = new boolean[chicken.size()];
		
		backTracking(0, 0);
		
		System.out.println(answer);
		
		
	}

}

File: 백준/Gold/15686. 치킨 배달/치킨 배달.py
import sys
input = sys.stdin.readline



def comb(n,r,start):
    global ans

    if n==M:
        result = 0
        for l in list(zip(*dist)):
            result += min(l)
        ans = result
        return


    if r==0:
        result = 0
        for l in list(zip(*select)):
            result += min(l)
        ans = result if ans > result else ans
        return

    for i in range(start,n):
        select.append(dist[i])
        comb(n,r-1,i+1)
        select.pop()



def solve():

    for i in range(len(chicken)):
        cr,cc = chicken[i]
        for j in range(len(house)):
            hr,hc = house[j]
            dist[i][j] = abs(cr-hr) + abs(cc-hc)


N,M = map(int,input().split())

city = []
house=[]
chicken=[]
ans = 1e9
for _ in range(N):
    city.append(list(map(int,input().split())))

for i in range(N):
    for j in range(N):
        if city[i][j]==1: house.append([i,j])
        elif city[i][j]==2: chicken.append([i,j])

dist=[[0]*len(house) for _ in range(len(chicken))]

solve()

select=[]
comb(len(chicken),M,0)

print(ans)
